packages/ejson/lib/src/encoding.dart

```diff
@@ -0,0 +1,223 @@
+
+EJsonValue _encodeCustom(Object value) {
+  final encoder = customEncoders[value.runtimeType.base];
+  if (encoder == null) {
+    throw MissingEncoder(value);
+  }
+  return encoder(value);
+}
+
+EJsonValue _encodeDate(DateTime value) {
+  return switch (relaxed) {
+    true => {'\$date': value.toIso8601String()},
+    false => {
+        '\$date': {'\$numberLong': value.millisecondsSinceEpoch.toString()},
+      },
+  };
+}
+
+EJsonValue _encodeDefined(Defined<dynamic> defined) => toEJson(defined.value);
+
+EJsonValue _encodeDocument(Map<dynamic, dynamic> map) => map.map((k, v) => MapEntry(k, toEJson(v)));
+
+EJsonValue _encodeDouble(double value) {
+  if (value.isNaN) {
+    return {'\$numberDouble': 'NaN'};
+  }
+  return switch (value) {
+    double.infinity => {'\$numberDouble': 'Infinity'},
+    double.negativeInfinity => {'\$numberDouble': '-Infinity'},
+    _ => switch (relaxed) {
+        true => value,
+        false => {'\$numberDouble': '$value'},
+      }
+  };
+}
+
+EJsonValue _encodeInt(int value, {bool long = true}) {
+  return switch (relaxed) {
+    true => value,
+    false => {'\$number${long ? 'Long' : 'Int'}': '$value'},
+  };
+}
+
+EJsonValue _encodeKey(Key key) => {'\$${key.name}Key': 1};
+
+@pragma('vm:prefer-inline')
+EJsonValue _encodeString(String value) => value;
+
+EJsonValue _encodeSymbol(Symbol value) => {'\$symbol': value.name};
+
+EJsonValue _encodeUndefined(Undefined<dynamic> undefined) => {'\$undefined': 1};
+
+EJsonValue _encodeUuid(Uuid uuid) => _encodeBinary(uuid.bytes.asUint8List(), "04");
+
+EJsonValue _encodeBinary(Uint8List buffer, String subtype) => {
+      '\$binary': {
+        'base64': base64.encode(buffer),
+        'subType': subtype,
+      },
+    };
+
+EJsonValue _encodeObjectId(ObjectId objectId) => {'\$oid': objectId.hexString};
+
+/// Exception thrown when no encoder is registered for the type of a [value].
+class MissingEncoder implements Exception {
+  final Object value;
+
+  MissingEncoder(this.value);
+
+  @override
+  String toString() => 'Missing encoder for type ${value.runtimeType} ($value)';
+}
+
+extension BoolEJsonEncoderExtension on bool {
+  /// Converts this [bool] to EJson
+  @pragma('vm:prefer-inline')
+  EJsonValue toEJson() => _encodeBool(this);
+}
+
+extension DateTimeEJsonEncoderExtension on DateTime {
+  /// Converts this [DateTime] to EJson
+  @pragma('vm:prefer-inline')
+  EJsonValue toEJson() => _encodeDate(this);
+}
+
+extension DefinedEJsonEncoderExtension on Defined<dynamic> {
+  /// Converts this [Defined] to EJson
+  @pragma('vm:prefer-inline')
+  EJsonValue toEJson() => _encodeDefined(this);
+}
+
+extension DoubleEJsonEncoderExtension on double {
+  /// Converts this [double] to EJson
+  @pragma('vm:prefer-inline')
+  EJsonValue toEJson() => _encodeDouble(this);
+}
+
+extension IntEJsonEncoderExtension on int {
+  /// Converts this [int] to EJson
+  @pragma('vm:prefer-inline')
+  EJsonValue toEJson({bool long = true}) => _encodeInt(this, long: long);
```

Comment:
I think using `long` here might be a bit confusing. I would suggest either `encodeAsLong` or `encodeAsInt32 = false`.